Chart Drift has never run: provision MEMEX_DEPLOY_REPO_TOKEN + the secret-free per-env overlays + AKS runCommand rights

`Chart Drift` (`.github/workflows/chart-drift.yml`, daily `15 6 * * *`) has failed **every run since it was added** — 3 for 3:

| run | when | result |
|---|---|---|
| [32001510403](https://github.com/Systemorph/MeshWeaver/actions/runs/32001510403) | 2026-08-17 06:24Z | failure |
| [31931303015](https://github.com/Systemorph/MeshWeaver/actions/runs/31931303015) | 2026-08-16 06:23Z | failure |
| [31869293070](https://github.com/Systemorph/MeshWeaver/actions/runs/31869293070) | 2026-08-15 06:22Z | failure |

**This is the designed behaviour, not a defect in the workflow.** Its header says so explicitly, and AGENTS.md → *"A gate NEVER tests its own inputs"* is why: a drift check that quietly skips when a credential is missing would report "no drift" on no evidence — strictly worse than not existing, because it also tells you the cluster is fine. So `preflight` fails RED and names what to provision. Filing this so the red is tracked rather than normalised into background noise — **drift detection for `memex`, `memex-cloud` and `atioz` is currently not running at all.**

## What the run actually says

```
##[error]chart-drift cannot run — provision the following, then re-run:
  • secrets.MEMEX_DEPLOY_REPO_TOKEN  read-only PAT for Systemorph/Memex, where the per-env
    values.<env>.public.yaml overlays live (they must be committed there first, secret-free)
```

`AZURE_CLIENT_ID` / `AZURE_TENANT_ID` / `AZURE_SUBSCRIPTION_ID` are all present. `MEMEX_DEPLOY_REPO_TOKEN` is the only empty one.

## To make it green — three things, in this order

**1. Commit the secret-free half of each env's helm overlay to `Systemorph/Memex`.**

The workflow checks out `Systemorph/Memex` at `deploy-repo/` and reads
`deployments/aks/<ns>/values.<release>.public.yaml` for each of:

| namespace | release | expected file |
|---|---|---|
| `memex` | `memex` | `deployments/aks/memex/values.memex.public.yaml` |
| `memex-cloud` | `memexcloud` | `deployments/aks/memex-cloud/values.memexcloud.public.yaml` |
| `atioz` | `atioz` | `deployments/aks/atioz/values.atioz.public.yaml` |

Include everything under `config:`, `persistence:`, `keda:`, `probes:`, `ingress:` — **nothing under `secrets:`**. The check compares the ConfigMap and the pod spec; it never reads a secret value. `check-chart-drift.sh` deliberately refuses to render with chart defaults (comparing the cluster against a chart nobody deployed produces confident nonsense), which is why the overlays have to exist rather than being optional. Optional per-env `portal-patch.json` is picked up as `--expect-patch` if present.

**2. Add `MEMEX_DEPLOY_REPO_TOKEN` as a repo secret** — a fine-grained PAT scoped to `Systemorph/Memex` with **Contents: Read-only** and nothing else.

**3. Grant the CI identity AKS command access.** The app registration behind `secrets.AZURE_CLIENT_ID` currently only has ACR rights (that is all `main-cd.yml` needs). Chart Drift additionally needs `Microsoft.ContainerService/managedClusters/runCommand/action` on the cluster. Scoped to the cluster resource `memexaks-cluster` (RG `memex-aks-rg`):

- **Azure Kubernetes Service Cluster User Role**
- **Azure Kubernetes Service RBAC Reader**

The check only ever issues `kubectl get`, so read-only is sufficient.

## Why it matters

`check-chart-drift.sh` was written 2026-08-13 to detect exactly the class of failure that took `memex.meshweaver.cloud` down the following day — and it had no caller at all until this workflow. Every day it stays unprovisioned is a day cluster-vs-chart divergence goes undetected. Related: the hand-applied config that `helm upgrade` silently reverts is precisely what this catches.

## Verifying

After all three land: `gh workflow run chart-drift.yml --ref main`. A completed run reports drift, if any, as red with each divergence classified `CLUSTER-ONLY` / `CHART-ONLY` / `DIFFERS`; green means the three namespaces match their charts.

## The premise has moved twice — correcting the record

### 1. The token IS provisioned, and the blocker is different

`secrets.MEMEX_DEPLOY_REPO_TOKEN` was added on **2026-08-17T11:12:42Z**. This issue's title still says "provision" it, which is no longer the ask.

What the preflight actually reports, on today's run (32454219730) and every run since 08-17:

```
##[error] secrets.MEMEX_DEPLOY_REPO_TOKEN is set but GitHub rejects it (HTTP 401, Bad credentials).
```

So it is **set but invalid** — revoked, expired, or truncated on copy. Worth noting the preflight is behaving exactly as designed: #1774 taught it to assert the token is **VALID, not merely present**, precisely so this state could not read as green. It is working.

### 2. …but the maintainer's direction is not to re-mint it at all

> *"there is such a token in memex.systemorph.com. all deploys should happen through there"*

That is correct, and verified:

- **`memex.systemorph.com` holds a `Deployments` space** that GitSyncs from **`Systemorph/Memex`** — the private deploy repo — subdirectory `mesh/Deployments`, `lastSyncedAt: 2026-08-19T20:06:37Z`. It is reading that repo successfully **right now**, with a credential that works.
- That space is already the deployment control plane: `Hosting/Deployment` records for `memex`, `memex-cloud` and `atioz`, plus `Hosting/BackupStore` and `Hosting/InstanceAction`.

So the instance that chart-drift needs data from **already has the credential and the synced content**. Minting a second PAT into GitHub Actions secrets duplicates a credential that exists, and re-creates the class of problem #1835/#1844 is retiring elsewhere:

> a consumer that needed **one package** asked for a standing GitHub credential to a **whole repository**, because there was no smaller thing to ask for.

Chart-drift needs the per-env `values.<env>.public.yaml` overlays. Those are in `Systemorph/Memex`, which the `Deployments` space already syncs.

### Proposed re-scope

**Retitle to "Chart Drift reads chart state from the Deployments space, not from a repo PAT."** The work becomes:

1. Chart-drift obtains overlay/chart state from `memex.systemorph.com`'s `Deployments` space rather than checking out `Systemorph/Memex`.
2. `MEMEX_DEPLOY_REPO_TOKEN` is then **removed**, not re-minted — one fewer standing credential, and no expiry to rediscover in three months.
3. The preflight keeps asserting whatever the new path needs, with the same fail-RED-naming-it contract.

Two things unchanged and still required either way: the AKS `runCommand` rights (the header notes the existing app registration already holds them, no role assignment needed), and the secret-free per-env overlays being committed to the deploy repo first.

**Open question for the maintainer:** whether chart-drift reaches the `Deployments` space over the MCP/HTTP surface, or whether the drift check should move *into* the mesh entirely and become a `Hosting/` capability beside `Hosting/Deployment` — which would fit the instance-lifecycle work already landing there.

## Correcting this issue: the secret is provisioned. It has never been valid.

The title and step 2 of this issue are wrong, and have been for eight days. Measured, not inferred:

```
$ gh api repos/Systemorph/MeshWeaver/actions/secrets/MEMEX_DEPLOY_REPO_TOKEN
{"name":"MEMEX_DEPLOY_REPO_TOKEN",
 "created_at":"2026-08-17T11:12:42Z",
 "updated_at":"2026-08-17T11:12:42Z"}     ← created_at == updated_at: never re-set since
```

And the preflight log of the most recent run (`32816762564`, 2026-08-25) — note it gets **past** the presence check, so `AZURE_*` and the token are all non-empty:

```
All external inputs are present.
##[error]secrets.MEMEX_DEPLOY_REPO_TOKEN is set but GitHub rejects it (HTTP 401, Bad credentials).
```

Every run since the secret was stored has ended there: the `workflow_dispatch` eleven seconds after it landed (`32023855740`, 11:12:53Z), then **nine consecutive schedules** 08-18 → 08-25. Twelve failures total, and the first three (08-15/16/17-06:24, quoted in this issue) are the only ones that were the originally-diagnosed "unprovisioned".

## So the remaining action is RE-MINT, not provision

**401 rules out the two adjacent explanations by construction**, which is why the probe distinguishes them:

| code | means | ruled out? |
|---|---|---|
| **401** | the value does not authenticate as anything | ← **this is what we get** |
| 403 | valid token, org owner has not approved it | yes, we'd see 403 |
| 404 | valid token, repo not in its scope / no Contents:Read | yes, we'd see 404 |

What is left is a value that is not a token: revoked, expired, truncated on copy, or never the token at all — a clipboard-piped `gh secret set` stores whatever was on the clipboard. **A trailing newline does this too** (`gh secret set NAME < token.txt`), goes straight into the `Authorization` header, and is invisible in the Actions UI because every secret renders as `***`.

## Status of the three steps

| step | actual state |
|---|---|
| 1. overlays committed to `Systemorph/Memex` | ✅ done 2026-08-17 (`2cd3eee`) — header item 2 |
| 2. `MEMEX_DEPLOY_REPO_TOKEN` | ⚠️ **provisioned 2026-08-17, rejected 401 ever since — needs a re-mint** |
| 3. AKS `runCommand` rights | ✅ already held; the item was written from an assumption that turned out to be wrong (the app registration has *Cluster Admin Role*) |

Do not read the daily red as "1 and 3 are outstanding too" — preflight never gets far enough to exercise them.

## What #2278 changes, and what it does not

Nothing about the workflow's shape: the preflight is correct as designed and stays exactly as strict. Two things:

- **The header stops lying.** Item 3 said `[OUTSTANDING]`, which reads as "not provisioned" and is what kept this misdiagnosed.
- **The 401 branch now says what is actually stored** — a length, a prefix *class* (`github_pat_` / `ghp_` / `gho_`-family / none), and a whitespace check. All derived; the value itself is never printed and never narrowed. Verified firing on a real trailing newline (`38 characters` where the clean token is 37, plus the named `printf %s | gh secret set` fix).

**Chart Drift stays RED until someone re-mints the PAT** — that is one manual action, and it is the whole remaining scope of this issue. It will not go green on a workflow edit, and the first completing run is expected to be legitimately very red (a local run found 121 divergences across 149 compared fields — the accumulated `kubectl patch` backlog, which is the finding, not a defect in the gate).

Re-titling suggestion: *"Chart Drift: MEMEX_DEPLOY_REPO_TOKEN authenticates as nothing (HTTP 401) — re-mint it"*.

--- a/.github/workflows/chart-drift.yml
+++ b/.github/workflows/chart-drift.yml
@@ -48,15 +48,25 @@ name: Chart Drift
 #
 #      Regenerate after any deliberate helm change, or the gate compares against a stale intent.
 #
-#   3. [OUTSTANDING] A fine-grained read-only PAT for Systemorph/Memex (Contents: Read-only, nothing
-#      else) added as secrets.MEMEX_DEPLOY_REPO_TOKEN. A PAT cannot be minted through the API or
-#      CLI, so this step is necessarily manual. Do NOT substitute the existing broad `GH_PAT` (it
-#      widens what this job can reach for no benefit), and do NOT move the overlays into this public
-#      repo to dodge the token — that would publish the deployment inventory.
+#   3. [DONE] Read access to Systemorph/Memex, as a GitHub App installation token minted per run
+#      (secrets.MESHWEAVER_APP_ID + MESHWEAVER_APP_PRIVATE_KEY, the org's `meshweaver-cloud` App).
 #
-# Until 3 lands, every run fails with the list above and nobody is misled about whether the cluster
-# matches the chart. Drift found by a run that DOES complete is reported the same way: red, with
-# each divergence classified CLUSTER-ONLY / CHART-ONLY / DIFFERS.
+#      🚨 This step used to read "a fine-grained read-only PAT ... a PAT cannot be minted through
+#      the API or CLI, so this step is necessarily manual". The first half is true and the second
+#      does not follow: an App INSTALLATION token is mintable, and the org already had an App
+#      installed with access to this repo. The PAT route was tried and it rotted precisely as PATs
+#      do — provisioned 2026-08-17, HTTP 401 "Bad credentials" on all nine runs after, so the check
+#      never ran once (#1709). A stored PAT has no owner, no expiry anyone watches, and fails in a
+#      way indistinguishable from a scope problem. The installation token is minted fresh each run
+#      and scoped to `repositories: Memex` with `permission-contents: read`, which is narrower than
+#      the PAT it replaces. Rotation has ONE home: Key Vault `Systemorph/github-app-privatekey`.
+#
+#      Still true, and still worth saying: do NOT substitute the broad `GH_PAT` (it widens what this
+#      job can reach for no benefit), and do NOT move the overlays into this public repo to dodge
+#      the credential — that would publish the deployment inventory.
+#
+# Drift found by a run that completes is reported red, with each divergence classified
+# CLUSTER-ONLY / CHART-ONLY / DIFFERS.
 #
 # 🔒 WHICH ENVIRONMENTS EXIST IS NOT PUBLIC. This workflow used to carry the namespace list and the
 # cluster/resource-group names inline — deployment identities, in a public repo, describing an
@@ -101,13 +111,15 @@ jobs:
           AZURE_CLIENT_ID: ${{ secrets.AZURE_CLIENT_ID }}
           AZURE_TENANT_ID: ${{ secrets.AZURE_TENANT_ID }}
           AZURE_SUBSCRIPTION_ID: ${{ secrets.AZURE_SUBSCRIPTION_ID }}
-          MEMEX_DEPLOY_REPO_TOKEN: ${{ secrets.MEMEX_DEPLOY_REPO_TOKEN }}
+          MESHWEAVER_APP_ID: ${{ secrets.MESHWEAVER_APP_ID }}
+          MESHWEAVER_APP_PRIVATE_KEY: ${{ secrets.MESHWEAVER_APP_PRIVATE_KEY }}
         run: |
           missing=()
           [ -n "$AZURE_CLIENT_ID" ]       || missing+=("secrets.AZURE_CLIENT_ID        OIDC app registration for azure/login")
           [ -n "$AZURE_TENANT_ID" ]       || missing+=("secrets.AZURE_TENANT_ID        the Systemorph AAD tenant")
           [ -n "$AZURE_SUBSCRIPTION_ID" ] || missing+=("secrets.AZURE_SUBSCRIPTION_ID  the subscription holding the AKS cluster")
-          [ -n "$MEMEX_DEPLOY_REPO_TOKEN" ] || missing+=("secrets.MEMEX_DEPLOY_REPO_TOKEN  read-only PAT for $DEPLOY_REPO, where the per-env values.<env>.public.yaml overlays live (see this file's header — they must be committed there first, secret-free)")
+          [ -n "$MESHWEAVER_APP_ID" ]     || missing+=("secrets.MESHWEAVER_APP_ID          app id of the org GitHub App that reads $DEPLOY_REPO, where the per-env values.<env>.public.yaml overlays live")
+          [ -n "$MESHWEAVER_APP_PRIVATE_KEY" ] || missing+=("secrets.MESHWEAVER_APP_PRIVATE_KEY  that App's private key (PEM); source of truth is Key Vault Systemorph/github-app-privatekey")
           if [ ${#missing[@]} -gt 0 ]; then
             echo "::error::chart-drift cannot run — provision the following, then re-run:"
             for m in "${missing[@]}"; do echo "  • $m"; done
@@ -127,9 +139,23 @@ jobs:
       # the real API, and name what each rejection means, because the three causes are
       # indistinguishable from the checkout error and take very different fixes. Read-only, one
       # request, no `if:`, no continue-on-error: it fails RED like everything else here.
+      # Cross-repo reads are a GitHub App installation token minted per run, never a stored PAT.
+      # A PAT was the original design and it rotted exactly as PATs do: provisioned 2026-08-17,
+      # HTTP 401 on every run after (#1709). An installation token cannot rot — it is minted fresh
+      # here, expires in an hour, and is scoped to the one repository named below.
+      - name: Mint an installation token for the deploy repo
+        id: deploy-token
+        uses: actions/create-github-app-token@v2
+        with:
+          app-id: ${{ secrets.MESHWEAVER_APP_ID }}
+          private-key: ${{ secrets.MESHWEAVER_APP_PRIVATE_KEY }}
+          owner: Systemorph
+          repositories: Memex
+          permission-contents: read
+
       - name: The token can actually READ the deploy repo
         env:
-          MEMEX_DEPLOY_REPO_TOKEN: ${{ secrets.MEMEX_DEPLOY_REPO_TOKEN }}
+          DEPLOY_TOKEN: ${{ steps.deploy-token.outputs.token }}
         run: |
           # 🚨 `|| code=000` is not defensive noise. `run:` steps are `bash -e`, so a curl that
           # fails at the TRANSPORT level (DNS, TLS, connection refused) makes this assignment
@@ -141,32 +167,31 @@ jobs:
           # answer into a 20-minute one.
           code=$(curl -sS -o /dev/null -w '%{http_code}' \
             --connect-timeout 10 --max-time 30 \
-            -H "Authorization: Bearer $MEMEX_DEPLOY_REPO_TOKEN" \
+            -H "Authorization: Bearer $DEPLOY_TOKEN" \
             -H "Accept: application/vnd.github+json" \
             -H "X-GitHub-Api-Version: 2022-11-28" \
             "https://api.github.com/repos/$DEPLOY_REPO") || code=000
           case "$code" in
             200)
-              echo "secrets.MEMEX_DEPLOY_REPO_TOKEN reads $DEPLOY_REPO (HTTP 200)." ;;
+              echo "The minted installation token reads $DEPLOY_REPO (HTTP 200)." ;;
             401)
-              echo "::error::secrets.MEMEX_DEPLOY_REPO_TOKEN is set but GitHub rejects it (HTTP 401, Bad credentials)."
-              echo "  The stored value is not a usable token. Most often it is not the token at all —"
-              echo "  a clipboard-piped 'gh secret set' stores whatever was on the clipboard — or the"
-              echo "  token was revoked, expired, or truncated on copy. Re-mint at"
-              echo "  https://github.com/settings/personal-access-tokens and store it again."
+              echo "::error::The minted installation token is rejected (HTTP 401, Bad credentials)."
+              echo "  The mint step succeeded, so the App id and key are a matching pair — but the"
+              echo "  token GitHub returned is not being accepted. This is the one outcome that"
+              echo "  should be impossible here; treat it as a GitHub-side fault and re-run. If it"
+              echo "  persists, confirm the App has not been deleted out from under the install."
               exit 1 ;;
             403)
-              echo "::error::secrets.MEMEX_DEPLOY_REPO_TOKEN is refused (HTTP 403)."
-              echo "  The token exists but is not authorised for $DEPLOY_REPO's organisation. A"
-              echo "  fine-grained PAT whose resource owner is an org stays inert until an org owner"
-              echo "  APPROVES it — check https://github.com/settings/personal-access-tokens for a"
-              echo "  'Pending owner approval' badge, and the org's pending-requests page to grant it."
+              echo "::error::The minted installation token is refused (HTTP 403)."
+              echo "  The App is installed but suspended, or its installation permissions were"
+              echo "  narrowed below Contents: Read. Check the App's installation on the Systemorph"
+              echo "  org and re-approve the permissions it requests."
               exit 1 ;;
             404)
-              echo "::error::secrets.MEMEX_DEPLOY_REPO_TOKEN authenticates but cannot see $DEPLOY_REPO (HTTP 404)."
+              echo "::error::The token authenticates but cannot see $DEPLOY_REPO (HTTP 404)."
               echo "  GitHub returns 404 rather than 403 for a repository a token may not read, so this"
-              echo "  is a SCOPE problem: the PAT's repository access does not include $DEPLOY_REPO, or"
-              echo "  it lacks 'Contents: Read-only'. Re-scope it — no other permission is needed."
+              echo "  is a SCOPE problem: the App's installation no longer includes $DEPLOY_REPO. Add"
+              echo "  the repository to the installation — no other permission is needed."
               exit 1 ;;
             000)
               echo "::error::Could not reach api.github.com to probe $DEPLOY_REPO (no HTTP response)."
@@ -181,7 +206,7 @@ jobs:
               echo "  re-run when the incident clears. Failing closed: indeterminate is not a pass."
               exit 1 ;;
             *)
-              echo "::error::Unexpected HTTP $code probing $DEPLOY_REPO with secrets.MEMEX_DEPLOY_REPO_TOKEN."
+              echo "::error::Unexpected HTTP $code probing $DEPLOY_REPO with the minted token."
               echo "  Treating as FAILURE: an indeterminate answer must never read as a working token."
               exit 1 ;;
           esac
@@ -195,11 +220,24 @@ jobs:
       aks_cluster: ${{ steps.envs.outputs.aks_cluster }}
       aks_resource_group: ${{ steps.envs.outputs.aks_resource_group }}
     steps:
+      # Cross-repo reads are a GitHub App installation token minted per run, never a stored PAT.
+      # A PAT was the original design and it rotted exactly as PATs do: provisioned 2026-08-17,
+      # HTTP 401 on every run after (#1709). An installation token cannot rot — it is minted fresh
+      # here, expires in an hour, and is scoped to the one repository named below.
+      - name: Mint an installation token for the deploy repo
+        id: deploy-token
+        uses: actions/create-github-app-token@v2
+        with:
+          app-id: ${{ secrets.MESHWEAVER_APP_ID }}
+          private-key: ${{ secrets.MESHWEAVER_APP_PRIVATE_KEY }}
+          owner: Systemorph
+          repositories: Memex
+          permission-contents: read
       - name: Check out the deployment repo
         uses: actions/checkout@v7
         with:
           repository: ${{ env.DEPLOY_REPO }}
-          token: ${{ secrets.MEMEX_DEPLOY_REPO_TOKEN }}
+          token: ${{ steps.deploy-token.outputs.token }}
           path: deploy-repo
       # No `if:`, no `continue-on-error:` — same rule as preflight, and for a sharper reason. An
       # empty or unreadable manifest yields an empty matrix, and an empty matrix SKIPS every drift
@@ -264,11 +302,24 @@ jobs:
         env: ${{ fromJson(needs.discover.outputs.matrix) }}
     steps:
       - uses: actions/checkout@v7
+      # Cross-repo reads are a GitHub App installation token minted per run, never a stored PAT.
+      # A PAT was the original design and it rotted exactly as PATs do: provisioned 2026-08-17,
+      # HTTP 401 on every run after (#1709). An installation token cannot rot — it is minted fresh
+      # here, expires in an hour, and is scoped to the one repository named below.
+      - name: Mint an installation token for the deploy repo
+        id: deploy-token
+        uses: actions/create-github-app-token@v2
+        with:
+          app-id: ${{ secrets.MESHWEAVER_APP_ID }}
+          private-key: ${{ secrets.MESHWEAVER_APP_PRIVATE_KEY }}
+          owner: Systemorph
+          repositories: Memex
+          permission-contents: read
       - name: Check out the per-env overlays
         uses: actions/checkout@v7
         with:
           repository: ${{ env.DEPLOY_REPO }}
-          token: ${{ secrets.MEMEX_DEPLOY_REPO_TOKEN }}
+          token: ${{ steps.deploy-token.outputs.token }}
           path: deploy-repo
       - name: Install helm
         uses: azure/setup-helm@v5

